Created cart drawer

heshvab opened the pull request:
- Added drawer context. Initially I thought to use modal provider for it but I I think separate providers are better in perspective in case we will have dialog that opens from inside of a drawer. Besides, separation makes code less complicated. 
- Added cart drawer. For now with mocked data.
- Moved modal provider and drawer provider to RootLayout. I did it because previously it was located higher in component tree than RouterProvider and react-router's links didn't work inside drawer. 
- Changed Product card a little so drawer opens after clicking Add to cart.
- I am going to create tasks for tests for cart drawer.

https://github.com/idx-academy/spa-orders/assets/62070431/a037a46d-17dd-4878-b584-c170f0118e40

heshvab committed:
Created cart drawer

heshvab committed:
Fixed import

heshvab committed:
Fixed mistakes

mhevyk reviewed (commented):
Ii my point of view animation of showing a drawer is too fast. There is an [API for drawer](https://mui.com/material-ui/api/drawer/), you can configure `transitionDuration`

mhevyk commented:
And will tests for `DrawerContext` be added within this PR?

heshvab commented:
> And will tests for `DrawerContext` be added within this PR?

I guess no. I thought to add them after adding items to cart will be implemented, together with some other tests related to cart

D1nviz commented on code/src/layouts/cart-drawer/CartDrawer.scss:
```diff
@@ -0,0 +1,50 @@
+  width: 400px;
+  display: flex;
+  flex-direction: column;
+  height: 100svh;
```
use the `$screen-height` instead

heorhiiss replied:
Thanks, I forgot we have such variable despited added it myself) fixed

D1nviz commented on code/src/layouts/cart-drawer/CartDrawer.scss:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  &__title {
+    font-size: $fs-lg;
```
I suggest setting `subtitle2` (it uses 20px size)  variant to the `AppTypography` and if you need bold for this text, you can use `fontWeight` prop which includes `extra-bold` variant. For semantic you can use `h2` component

heorhiiss replied:
Thanks, I used suggested approach in this case

D1nviz commented on code/src/layouts/cart-drawer/message/index.ts:
```diff
@@ -0,0 +1,4 @@
+import en from "./en.json";
+import uk from "./uk.json";
```
use alias for imports

heorhiiss replied:
Thanks, I fixed it

OliaHrytsak commented on code/src/components/product-card/ProductCard.tsx:
```diff
@@ -4,45 +4,45 @@ import AppBox from "@/components/app-box/AppBox";
       </AppBox>
-    </AppLink>
+      <AppBox>
+        <AppTypography component={AppLink} to="/" variant="caption">
```
Why are you decided to move link there? Maybe it would be better user experience to wrap into link img+name?

heorhiiss replied:
We have button `Add to cart` inside of a `ProductCard` and if link wraps entire `ProductCard` - by clicking on a button we also click on a link, so user is navigated which is not desired behaviour when user clicks on a `Add to card` button. Probably it can be solved with `stopPropagation`, but that is in general a bad practice to have button inside of a link. So for now I made product title as a link.

OliaHrytsak replied:
I mean img+title , not all card

heorhiiss replied:
Sorry for misunderstanding, I added link around img and title

OliaHrytsak commented on code/src/context/DrawerContext.tsx:
```diff
@@ -0,0 +1,67 @@
+    setIsOpen(false);
+  };
+
+  const toggleDrawer = (component: ReactElement) => {
```
maybe we can awoid this if/else ?

`drawer ? closeDrawer() : openDrawer(component);`

heorhiiss replied:
Thanks for noticing, fixed

heshvab committed:
Fixed comments

heshvab commented:
> Ii my point of view animation of showing a drawer is too fast. There is an [API for drawer](https://mui.com/material-ui/api/drawer/), you can configure `transitionDuration`
I slowed animation speed to 400ms (was 300ms). Could you take a look? I think that is optimal value because it is not too fast and not too slow at the same time.

OliaHrytsak commented on code/src/layouts/cart-drawer/cart-drawer-item/CartDrawerItem.scss:
```diff
@@ -0,0 +1,49 @@
+@import "@design-system";
+
+.cart-item {
```
Don't forget about spa prefix

heorhiiss replied:
If I remember correctly, we use `spa` prefix only for components, prefixed with `App`?

heshvab committed:
Added link around img and title

mhevyk commented:
![image](https://github.com/idx-academy/spa-orders/assets/109922101/d9a584ac-78d3-446e-97b1-391378d8908e)

1. This button is full height. I think it would be better to make it smaller.
2. Please add `object-fit: cover`-like staff, because images from backend can differ and it will cause ui bugs
3. Add `text-truncate` mixin for name of product, because name can be very long and it would be better to make our layout more stable and independent of content we get from the server

heshvab committed:
Fixed button, text and image styles

heshvab commented:
> ![image](https://private-user-images.githubusercontent.com/109922101/347318048-d9a584ac-78d3-446e-97b1-391378d8908e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1OTk2OTgsIm5iZiI6MTcyMDU5OTM5OCwicGF0aCI6Ii8xMDk5MjIxMDEvMzQ3MzE4MDQ4LWQ5YTU4NGFjLTc4ZDMtNDQ2ZS05N2IxLTM5MTM3OGQ4OTA4ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMFQwODE2MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MWVlYzZlYjE5Y2Q3MDNiM2IzMjNlNTgyOWJmYTBkNWMzMGQwYzAwYmIyOGU4MDVjZmUwOTc5ZWU5Zjk0ZWZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.78e0VkwyVg0D9ClvcLTCkVsjisBVD4-Hbso65nljSKs)
> 
> 1. This button is full height. I think it would be better to make it smaller.
> 2. Please add `object-fit: cover`-like staff, because images from backend can differ and it will cause ui bugs
> 3. Add `text-truncate` mixin for name of product, because name can be very long and it would be better to make our layout more stable and independent of content we get from the server

Changed button
Added object fit and tested it on different images
Added text truncate and tested it

![image](https://github.com/idx-academy/spa-orders/assets/62070431/10c6addf-bf3b-493c-a13a-eb89fa8740a2)
